Store the customer currencies data as an option

eduardoumpierre opened the pull request:
Fixes #7000

#### Changes proposed in this Pull Request

This PR aims to improve the `register_customer_currencies` performance by replacing the expensive calculation with an option that is updated on every new order. The expensive `get_all_customer_currencies` call will be called only the first time, and then its value will be used to set an option.

#### Benchmark

The benchmark was done using `microtime()` to calculate the start and end times of the function execution.

##### Benchmark with 205 orders

<table>
    <thead>
        <tr>
            <th>#</th>
            <th>Previous</th>
            <th>Current</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td><code>0.001592874526977539μs</code></td>
            <td><code>0.000027179718017578125μs</code></td>
        </tr>
        <tr>
            <td>2</td>
            <td><code>0.0019979476928710938μs</code></td>
            <td><code>0.000025987625122070312μs</code></td>
        </tr>
       <tr>
            <td>3</td>
            <td><code>0.0030279159545898438μs</code></td>
            <td><code>0.000030040740966796875μs</code></td>
        </tr>
       <tr>
            <td>4</td>
            <td><code>0.0017621517181396484μs</code></td>
            <td><code>0.000024080276489257812μs</code></td>
        </tr>
       <tr>
            <td>5</td>
            <td><code>0.00202178955078125μs</code></td>
            <td><code>0.000032901763916015625μs</code></td>
        </tr>
        <tr>
            <td>Avg</td>
            <td><code>0.002080535888672μs</code></td>
            <td><code>0.000028038024902μs</code></td>
        </tr>
        <tr>
            <td>Diff</td>
            <td colspan="2">~74x speedup</td>
        </tr>
    </tbody>
</table>

##### Benchmark with 5723 orders

<table>
    <thead>
        <tr>
            <th>#</th>
            <th>Previous</th>
            <th>Current</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td><code>0.5904631614685059μs</code></td>
            <td><code>0.00003600120544433594μs</code></td>
        </tr>
        <tr>
            <td>2</td>
            <td><code>0.6487610340118408μs</code></td>
            <td><code>0.00003600120544433594μs</code></td>
        </tr>
       <tr>
            <td>3</td>
            <td><code>0.8029019832611084μs</code></td>
            <td><code>0.000032901763916015625μs</code></td>
        </tr>
       <tr>
            <td>4</td>
            <td><code>0.6125071048736572μs</code></td>
            <td><code>0.00003409385681152344μs</code></td>
        </tr>
       <tr>
            <td>5</td>
            <td><code>0.5736191272735596μs</code></td>
            <td><code>0.000031948089599609375μs</code></td>
        </tr>
        <tr>
            <td>Avg</td>
            <td><code>0.642</code></td>
            <td><code>0.00003416</code></td>
        </tr>
        <tr>
            <td>Diff</td>
            <td colspan="2">~18,793x speedup</td>
        </tr>
    </tbody>
</table>

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout to this branch
* Go to the analytics page `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
  * The new option will be set on this page load
* After the page loads, open a new tab on PhpMyAdmin `http://localhost:8083/` and run the following query:
```sql
SELECT * FROM `wp_options` WHERE `option_name` LIKE "wcpay_multi_currency_stored_customer_currencies";
```
* Go to the Multi-currency settings page `/wp-admin/admin.php?page=wc-settings&tab=wcpay_multi_currency` and enable a new currency
* Go to your shop and perform the checkout using the currency you've just added
* Go to the PhpMyAdmin tab and refresh the query
* Check the array on `option_value`, it should contain the currency code
* Go to the analytics page `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
* Check the "Customer currency" select – it should also contain the currency code option

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Store customer currencies as an option

Store customer currencies to optimize performance and reduce expenses.

Maintain an updated array of currencies for every new order.

eduardoumpierre committed:
Update customer currencies option on intention request

eduardoumpierre committed:
Move option key constant to the MultiCurrency class

eduardoumpierre committed:
Update unit tests

eduardoumpierre committed:
Add changelog entry

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7143` or branch name `update/7000-improve-analytics-call-performance` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/7000-improve-analytics-call-performance) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3f37c4dc5f578523b39ed8827e9231d635b73d0c
- Build time: 2023-09-28 18:32:24 UTC

_Note: the build is updated when a new commit is pushed to this PR._

eduardoumpierre commented on includes/core/server/request/class-create-intention.php:
```diff
@@ -194,6 +194,8 @@ public function set_shipping( array $shipping ) {
 	 * @return mixed           Either the same response, or the correct object.
 	 */
 	public function format_response( $response ) {
+		$this->api_client->update_customer_currencies( $response );
```
Should we hook this into a `woocommerce_thankyou` action, for example?

eduardoumpierre replied:
I tested this action with different flows and found that it only works when the payment is completed (ha!), despite the documentation stating that it should also be triggered by the `on hold`, `pending`, `failed`, and `cancelled` statuses. I tested the capture later flow and the action was not triggered.

Since there is an advanced filter on the analytics page that allows the merchant to filter by any status, I believe we should update the customer currencies option even if the order is created manually.

<table>
<tr>
<td>
<img width="500" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/16882226/e02693cf-10a8-4385-8630-804ffb2deefe">
</td>
<td>
<img width="500" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/16882226/d8678350-33ad-4f06-b8aa-7db467654cd8">
</td>
</tr>
<tr>
<td>
<i>Advanced filters</i>
</td>
<td>
<i>Available statuses</i>
</td>
</tr>
</table>

After testing several available actions, I found that `woocommerce_order_status_changed` made the most sense to me. This is because there are multiple entry points to the payment flow, and the action also supports manually created orders.

<table>
<tr>
<td>
<img src="https://github.com/Automattic/woocommerce-payments/assets/16882226/8fa09424-d703-43fb-883b-eb171ce330de" />
</td>
</tr>
<tr>
<td>
<a href="https://woocommerce.com/document/managing-orders/#visual-diagram-to-illustrate-order-statuses"><i>Visual Diagram to Illustrate Order Statuses</i></a>
</td>
</tr>
</table>

It was updated on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/1b5bf554a57bf08b336d0fb62168143fbb15a207. Thanks for the suggestion, and let me know your thoughts on this!

eduardoumpierre commented:
I'll add additional unit tests for the `update_customer_currencies` method, but it would be good to receive some feedback while I work on it, since I'll be back next Monday.

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Move customer currencies option logic to the register method

eduardoumpierre committed:
Move the update customer currencies method to the MultiCurrency class

The method will be called by a WooCommerce hook.

eduardoumpierre commented on includes/multi-currency/Analytics.php:
```diff
@@ -115,11 +114,18 @@ public function register_admin_scripts() {
+		if ( ! $currencies ) {
+			$currencies = $this->multi_currency->get_all_customer_currencies();
+			update_option( MultiCurrency::CUSTOMER_CURRENCIES_KEY, $currencies );
+		}
```
It does make sense. I updated it on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/5dd5cd8b65a407c94d65a17edc989be87eee8f0f. Thanks for the suggestion!

eduardoumpierre committed:
Add unit tests for update customer currencies method

tpaksu commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -544,6 +548,39 @@ public function update_single_currency_settings( string $currency_code, string $
+
+		// Skip if the currency is not a string or if the option is invalid.
+		if ( ! is_string( $currency ) || ! is_array( $currencies ) ) {
+			return;
```
I'm thinking about the initial migration, let's assume I upgraded to the version that contains this change for the first time, I won't have the option set. `get_option` will return `false` here, right? So the customer currency won't be updated unless I go to the Analytics section to have it updated. What do you think about splitting the condition into two parts, one checking and bailing if the order currency isn't valid, and the second part resetting the currency option when the option doesn't exist, or isn't valid?

tpaksu replied:
Ok, missed the line in `register_customer_currencies` method on the MC class, the initialization will fill in that data. It was folded by GH, needed to unfold it :)

tpaksu replied:
Nope, it was on the `Analytics` initialization. So the question stays valid :)

eduardoumpierre replied:
> So the customer currency won't be updated unless I go to the Analytics section to have it updated.

The migration will happen on any page of the admin dashboard if MultiCurrency is enabled, not necessarily on the Analytics page, and most probably will happen right after the plugin update process. This is the path that will lead to this behavior:

1. The Init method is called on the [MultiCurrency constructor](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/MultiCurrency.php#L231)
2. The Analytics class [gets instantiated](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/MultiCurrency.php#L272)
3. The init method is called on the [Analytics constructor](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Analytics.php#L53)
4. The `register_customer_currencies` method is called [on the admin dashboard](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Analytics.php#L64)

This can be tested by removing the `wcpay_multi_currency_customer_currencies` option from the local database and then accessing any page on the admin dashboard – it should be generated again.

eduardoumpierre replied:
As Marcin commented, after having a slow first setup (for big merchants), it will be faster on the following page interactions.

Do you still have concerns after my comment above?

tpaksu replied:
That makes sense, thanks for the explanation! The customer currencies will be updated whenever the page loads with initializing the MultiCurrency class then. Gotcha.

tpaksu replied:
Oh, sorry, it is to be updated when the user is admin and has the rights to `manage_woocommerce`. Since we only use this data for the Analytics section, it won't matter, but when a buyer purchases something on the frontend, until the admin is logged in, the option will not get populated. And it will be a problem when someone tries to use that in the frontend on a fresh site. Has a low chance to happen though.

https://github.com/Automattic/woocommerce-payments/blob/5f75b8fd87ad8d34d350a7d7c0c4774fb60d6efe/includes/multi-currency/Analytics.php#L62-L65

Do you think we should cover all possibilities by replacing the `get_option` call with `self::get_all_customer_currencies` in https://github.com/Automattic/woocommerce-payments/blob/419ff2cbf599d1dd26eea966f9b18469f7a95a41/includes/multi-currency/MultiCurrency.php#L566 ? I think Marcin was also suggesting that.

eduardoumpierre replied:
That makes sense. I updated the code on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/b49072944d048424c7f34c236471096540fcfb3f. Thanks for the suggestion @tpaksu and @marcinbot!

tpaksu commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -1451,32 +1488,24 @@ public function is_multi_currency_settings_page(): bool {
+		}
+
+		if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) &&
+				\Automattic\WooCommerce\Utilities\OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			$currencies = $wpdb->get_col( "SELECT DISTINCT(currency) FROM {$wpdb->prefix}wc_orders" );
+		} else {
+			$currencies = $wpdb->get_col( "SELECT DISTINCT(meta_value) FROM {$wpdb->postmeta} WHERE meta_key = '_order_currency'" );
```
Added 6902 to merge queue now.

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Fix the customer currencies return format

eduardoumpierre committed:
Convert array values to uppercase before comparison

eduardoumpierre commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -544,6 +548,39 @@ public function update_single_currency_settings( string $currency_code, string $
+			return;
+		}
+
+		$currency = strtoupper( $currency );
```
Nice call! I updated the code to guarantee that the content of `currencies` is always uppercase on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/419ff2cbf599d1dd26eea966f9b18469f7a95a41.

eduardoumpierre committed:
Get the latest customer currencies data if option is not available

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Fix unit test

eduardoumpierre committed:
Merge get customer currencies methods

eduardoumpierre committed:
Transform array to uppercase before storing the customer currencies option

eduardoumpierre commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -1446,11 +1484,11 @@ public function is_multi_currency_settings_page(): bool {
 	 * @return array
 	 */
-	public function callable_get_customer_currencies() {
+	public function get_customer_currencies() {
```
Great point, I also wasn't happy with the naming of those two functions. I merged both on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/872060c6f4aa096fc0f8fa6ae7270c95c07c0124. Thanks for the suggestion!

eduardoumpierre commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -544,6 +548,40 @@ public function update_single_currency_settings( string $currency_code, string $
+		}
+
+		$currency   = strtoupper( $currency );
+		$currencies = array_map( 'strtoupper', $currencies );
```
It does make sense. I updated the code to transform the array before saving the option on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/c947e6ce5bcd008aa3da77f2cdf3d322acac10af. Thanks for the suggestion!

eduardoumpierre replied:
I thought about saving it directly as uppercase because the option wouldn't exist prior to this code execution, so it would always get the uppercase values. I do agree that it's better to guarantee that we're always dealing with uppercase currencies. The code was updated on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/862dcbeacd61d2ef2e46e506e6241a3f725d36fe.

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/7000-improve-analytics-call-performance

tpaksu commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -544,6 +548,39 @@ public function update_single_currency_settings( string $currency_code, string $
+		$currencies = self::get_all_customer_currencies();
+
+		// Skip if the currency is not a string or if the option is invalid.
+		if ( ! is_string( $currency ) || ! is_array( $currencies ) ) {
```
The `$currencies` variable doesn't have the chance to be anything else than an array now after switching to `self::get_all_customer_currencies()`, right? Maybe we should remove it. It being an empty array is also fine for the script to continue as far as I can see.

eduardoumpierre replied:
I agree with you. It also appears unnecessary to include the check for `! is_string( $currency )` since orders cannot be created with a non-string currency. Perhaps we should remove the check entirely?

eduardoumpierre replied:
I updated the code on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/71ce3fd515e45c3d8617138c916ec6d79f875c51. Thanks for the suggestion and let me know your thoughts on removing both checks.

eduardoumpierre committed:
Remove unnecessary currency checks

As the currency will always be a string, there is no point
on checking if it's not a string. Following the same idea,
after the latest changes on get_all_customer_currencies,
it always return an array, so it is not necessary to check
if $currencies is an array.

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Transform currencies to uppercase before returning

eduardoumpierre committed:
Update stored customer currencies key

eduardoumpierre commented on includes/multi-currency/MultiCurrency.php:
```diff
@@ -27,9 +27,10 @@
+	const CURRENCY_SESSION_KEY    = 'wcpay_currency';
+	const CURRENCY_META_KEY       = 'wcpay_currency';
+	const FILTER_PREFIX           = 'wcpay_multi_currency_';
+	const CUSTOMER_CURRENCIES_KEY = self::FILTER_PREFIX . 'customer_currencies';
```
Nice catch! I updated the key to be `wcpay_multi_currency_stored_customer_currencies` on https://github.com/Automattic/woocommerce-payments/pull/7143/commits/3d101abf27e55361640ffa933e4909f25b4e9112.

eduardoumpierre committed:
Merge branch 'develop' into update/7000-improve-analytics-call-performance

tpaksu reviewed (approved):
Code looks good, is understandable and has no problems, I had a small problem when running the SQL to get the updated field, then realized the option name was `wcpay_multi_currency_stored_customer_currencies`, not the one written in the query, now it's all good :) Thanks for working on this. LGTM!

eduardoumpierre committed:
Merge branch 'develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Merge branch 'develop' into update/7000-improve-analytics-call-performance

eduardoumpierre committed:
Merge branch 'develop' into update/7000-improve-analytics-call-performance
